specs/WebMessageObjects.md

```diff
@@ -0,0 +1,176 @@
+
+Currently the only supported DOM object type with this API is:
+- [File](https://developer.mozilla.org/en-US/docs/Web/API/File)
+
+Page content in WebView2 can pass objects to the app via the
+`chrome.webview.postMessageWithAdditionalObjects(string message, array<object> objects)` content API
+that takes in the array of such supported DOM objects or you can also use
+(ExecuteScript)[https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.executescriptasync]
+with same API. If an invalid or unsupported object is passed via this API, an exception will be
+thrown to content and the message will fail to post.
+
+On the WebMessageReceived event handler, the app will retrieve the native representation of objects
+via `AdditionalObjects` property and cast the passed objects to their native types. For example, the
+DOM File object will be exposed as a `CoreWebView2File` object, allowing the app to read the path of
+the passed DOM File object.
+
+# Examples
+## Read the File path of a DOM File
+
+Use this API with a DOM File object to be able to get the path of files dropped on WebView2. The
+HTML and JavaScript snippets are part of the page content code running inside WebView2.
+
+```html
+<!-- File upload location -->
+<input type="file" id="files" multiple />
+```
+
+```javascript
+const input = document.getElementById('files');
+input.addEventListener('change', function() {
+    const curFiles = input.files;
+    chrome.webview.postMessageWithAdditionalObjects("FilesDropped", curFiles);
+});
+```
+
+```cpp
+m_webView->add_WebMessageReceived(
+    Callback<ICoreWebView2WebMessageReceivedEventHandler>(
+        [this](ICoreWebView2* sender, ICoreWebView2WebMessageReceivedEventArgs* args)
+        {
+            wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs2> args2 =
+                wil::com_ptr<ICoreWebView2WebMessageReceivedEventArgs>(args)
+                    .query<ICoreWebView2WebMessageReceivedEventArgs2>();
+            wil::com_ptr<ICoreWebView2WebMessageObjectCollectionView>
+                objectsCollection;
+            args2->get_AdditionalObjects(&objectsCollection);
+            unsigned int length;
+            objectsCollection->get_Count(&length);
+            std::vector<std::wstring> paths;
+
+            for (unsigned int i = 0; i < length; i++)
+            {
+                wil::com_ptr<IUnknown> object;
+                objectsCollection->GetValueAtIndex(i, &object);
+
+                wil::com_ptr<ICoreWebView2File> file =
+                    object.query<ICoreWebView2File>();
+                if (file)
+                {
+                    // Add the file to message to be sent back to webview
+                    wil::unique_cotaskmem_string path;
+                    file->get_Path(&path);
+                    paths.push_back(path.get());
+                }
+            }
+
+            return S_OK;
```

Comment:
Would be nice to do something with the paths variable after filling it in. Even if its only adding a call to a method that you don't show the definition of.
```suggestion

            ProcessFiles(files);
            return S_OK;
```